Environment tests fail if $UPDATE_OS_ENVIRON = True

<!--- Provide a general summary of the issue in the Title above -->
<!--- If you have a question along the lines of "How do I do this Bash command in xonsh"
please first look over the Bash to Xonsh translation guide: https://xon.sh/bash_to_xsh.html
If you don't find an answer there, please do open an issue! -->

## xonfig

<details>

```
$ xonfig
+------------------+------------------------------------+
| xonsh            | 0.14.2                             |
| Python           | 3.12.0                             |
| PLY              | 3.11                               |
| have readline    | True                               |
| prompt toolkit   | 3.0.41                             |
| shell type       | prompt_toolkit                     |
| history backend  | json                               |
| pygments         | 2.17.2                             |
| on posix         | True                               |
| on linux         | False                              |
| on darwin        | True                               |
| on windows       | False                              |
| on cygwin        | False                              |
| on msys2         | False                              |
| is superuser     | False                              |
| default encoding | utf-8                              |
| xonsh encoding   | utf-8                              |
| encoding errors  | surrogateescape                    |
| xontrib 1        | term_integration                   |
| xontrib 2        | broot                              |
| xontrib 3        | abbrevs                            |
| xontrib 4        | direnv                             |
| RC file 1        | /Users/qyriad/.config/xonsh/rc.xsh |
+------------------+------------------------------------+
```

</details>

## Expected Behavior

Presumably tests should succeed regardless of various xonsh options, or alternatively tests shouldn't source xonshrc.

## Current Behavior
<!--- Tell us what happens instead of the expected behavior -->
<!--- If part of your bug report is a traceback, please first enter debug mode before triggering the error
To enter debug mode, set the environment variable `XONSH_DEBUG=1` _before_ starting `xonsh`.
On Linux and OSX, an easy way to to do this is to run `env XONSH_DEBUG=1 xonsh` -->

If `$UPDATE_OS_ENVIRON` is set to `True`, then `test_deregister_custom_var` fails with:
```python
    def test_deregister_custom_var():
        env = Env()

        env.register("MY_SPECIAL_VAR", type="env_path")
        env.deregister("MY_SPECIAL_VAR")
>       assert "MY_SPECIAL_VAR" not in env
E       AssertionError: assert 'MY_SPECIAL_VAR' not in xonsh.environ.Env(...)

tests/test_environ.py:495: AssertionError
========================================================================================================== short test summary info ===========================================================================================================
FAILED tests/test_environ.py::test_deregister_custom_var - AssertionError: assert 'MY_SPECIAL_VAR' not in xonsh.environ.Env(...)
```

If we modify the test a little:
```diff
diff --git a/tests/test_environ.py b/tests/test_environ.py
index 220bcec7..b59c1c91 100644
--- a/tests/test_environ.py
+++ b/tests/test_environ.py
@@ -490,6 +490,8 @@ def test_register_custom_var_env_path():
 def test_deregister_custom_var():
     env = Env()
 
+    assert "MY_SPECIAL_VAR" not in os.environ
+
     env.register("MY_SPECIAL_VAR", type="env_path")
     env.deregister("MY_SPECIAL_VAR")
     assert "MY_SPECIAL_VAR" not in env
```

We can see that the test is failing because `MY_SPECIAL_VAR` is already in the *OS* environment before the actual testing happens, causing the later deregister to not remove it from the environment:

```python
    def test_deregister_custom_var():
        env = Env()

>       assert "MY_SPECIAL_VAR" not in os.environ
E       AssertionError: assert 'MY_SPECIAL_VAR' not in environ({'BASH_COMPLETIONS': '/usr/local/share/bash-completion/bash_completion:/usr/local/etc/bash_completion:/opt/hom...home/wakka/bin', 'MY_PATH_VAR': '', 'PYTEST_CURRENT_TEST': 'tests/test_environ.py::test_deregister_custom_var (call)'})
E        +  where environ({'BASH_COMPLETIONS': '/usr/local/share/bash-completion/bash_completion:/usr/local/etc/bash_completion:/opt/hom...home/wakka/bin', 'MY_PATH_VAR': '', 'PYTEST_CURRENT_TEST': 'tests/test_environ.py::test_deregister_custom_var (call)'}) = os.environ

tests/test_environ.py:493: AssertionError
========================================================================================================== short test summary info ===========================================================================================================
FAILED tests/test_environ.py::test_deregister_custom_var - AssertionError: assert 'MY_SPECIAL_VAR' not in environ({'BASH_COMPLETIONS': '/usr/local/share/bash-completion/bash_completion:/usr/local/etc/bash_completion:/opt/hom...home/wakka/bin', 'MY_PATH_VAR': '', 'PYTEST_CURRENT_TEST': 'tests...
```

`MY_SPECIAL_VAR` is getting in there from the earlier tests; if we run that specific test by itself with `pytest tests/test_environ.py -k test_deregister_custom_var`, it succeeds just fine.


## Steps to Reproduce
<!--- Please try to write out a minimal reproducible snippet to trigger the bug, it will help us fix it! -->

1. Set `$UPDATE_OS_ENVIRON = True`, either in a parent shell, or in `xonshrc` (**note that the `direnv` xontrib sets this!**)
2. Run xonsh's tests with `pytest`, or simply run the affected tests with `pytest tests/test_environ.py`.
3. Observe that the test fails

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

--- a/tests/test_environ.py
+++ b/tests/test_environ.py
@@ -31,6 +31,12 @@ from xonsh.pytest.tools import skip_if_on_unix
 from xonsh.tools import DefaultNotGiven, always_true
 
 
+@pytest.fixture(autouse=True)
+def set_env(monkeypatch):
+    # To avoid tests intersection (#5236)
+    monkeypatch.setenv("UPDATE_OS_ENVIRON", "False")
+
+
 def test_env_normal():
     env = Env(VAR="wakka")
     assert "wakka" == env["VAR"]

